src/PasswordPolicy.php

```diff
@@ -0,0 +1,289 @@
+    $this->minimumRequiredConstraints = $minimum_required_constraints;
+    $this->constraints = $constraints;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createFromRole(RoleInterface $role): ?PasswordPolicyInterface {
+    if ($third_party_settings = $role->getThirdPartySettings('password_enhancements')) {
+      return new static($role,
+        $third_party_settings['expireSeconds'] ?? self::PASSWORD_NO_EXPIRY,
+        $third_party_settings['expireWarnSeconds'] ?? self::PASSWORD_NO_WARNING,
+        $third_party_settings['expiryWarningMessage'] ?? 'Your password will expire on @date_time, please <a href="@url">change your password</a> before it expires to prevent any potential data loss.',
+        $third_party_settings['minimumRequiredConstraints'] ?? 1,
+        $third_party_settings['constraints'] ?? []);
+    }
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function loadByRoles(array $roles): ?PasswordPolicyInterface {
+    $policies = \Drupal::service('password_enhancements.password_constraint')->loadMultiplePoliciesByRoles($roles);
+    return $policies ? reset($policies) : NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireDays(): int {
+    return floor($this->getExpireSeconds() / 86400);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireSeconds(): int {
+    return $this->expireSeconds;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireWarnSeconds(): int {
+    return $this->expireWarnSeconds;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpireWarnDays(): int {
+    return floor($this->getExpireWarnSeconds() / 86400);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpiryWarningMessage(): ?string {
+    return $this->expiryWarningMessage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getMinimumRequiredConstraints(): int {
+    return $this->minimumRequiredConstraints;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getPriority(): int {
+    return $this->role->getWeight();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRole(): RoleInterface {
+    return $this->role;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setExpireSeconds(int $seconds): PasswordPolicyInterface {
+    if ($seconds < 0) {
+      throw new PolicyInvalidArgumentException('The expiry given in seconds must be a positive integer or zero.');
+    }
+    $this->expireSeconds = $seconds;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setExpireWarnSeconds(int $seconds): PasswordPolicyInterface {
+    if ($seconds < 0) {
+      throw new PolicyInvalidArgumentException('The expiry warning in seconds must be a positive integer or zero.');
+    }
+    $this->expireWarnSeconds = $seconds;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setExpiryWarningMessage(?string $message): PasswordPolicyInterface {
```

Comment:
Based on the above explanation, I am not sure whether we should allow here to change the value of a possible string value to null again after an object has been built.

Reply:
There's no need to store the expiry warning message when it's never to be displayed. See above for more details.